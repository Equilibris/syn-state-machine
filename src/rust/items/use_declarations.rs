use crate::*;

materialize! {
    #[derive(Debug)]
    pub struct UseDeclaration {
        <- KwUse;
        tree <- UseTree;
        <- Semi;
    }
}

#[derive(Debug)]
pub enum UseTree {
    Star(Option<Option<SimplePath>>),
    Branch(Option<Option<SimplePath>>, InterlaceTrail<Box<Self>, Comma>),
    Simple(SimplePath, Option<IdentifierOrUnder>),
}

impl Parse for UseTree {
    fn parse<'a>(input: &mut ParseBuffer<'a>) -> Result<Self> {
        Ok(
            match input.parse::<Sum3<
                (
                    Option<(Option<SimplePath>, PeekAsParse<PathSep>)>,
                    PeekAsParse<Star>,
                ),
                (Option<(Option<SimplePath>, PeekAsParse<PathSep>)>, Brace<_>),
                (SimplePath, Option<(KwAs, _)>),
            >>()? {
                Sum3::V0(a) => Self::Star(a.0.map(|v| v.0)),
                Sum3::V1(a) => Self::Branch(a.0.map(|v| v.0), a.1 .0),
                Sum3::V2(a) => Self::Simple(a.0, a.1.map(|v| v.1)),
            },
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    insta_match_test!(+it_matches_simple_path, UseDeclaration : use hello::world; );
    insta_match_test!(+it_matches_simple_path_as, UseDeclaration : use hello::world as h; );
    insta_match_test!(+it_matches_star_path, UseDeclaration : use hello::*; );
    insta_match_test!(+it_matches_complex_path, UseDeclaration :  use { hello::*, world::hi as Hi, nested::{ hello::world, hi }, }; );
}
