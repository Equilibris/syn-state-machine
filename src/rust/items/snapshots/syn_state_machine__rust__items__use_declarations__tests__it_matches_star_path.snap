---
source: src/rust/items/use_declarations.rs
expression: "insta_match_test!(+ parse UseDeclaration : use hello :: * ;)"
---
Ok(
    UseDeclaration {
        tree: Star(
            Some(
                Some(
                    HigherOrderPath {
                        leading: false,
                        segments: MinLength(
                            Interlace {
                                values: [
                                    Identifier(
                                        Identifier(
                                            Ident(
                                                hello,
                                            ),
                                        ),
                                    ),
                                ],
                            },
                        ),
                    },
                ),
            ),
        ),
    },
)
